contractcourt/htlc_incoming_contest_resolver.go

```diff
@@ -57,8 +57,11 @@ func (h *htlcIncomingContestResolver) Resolve() (ContractResolver, error) {
 		return nil, err
 	}
 
-	// If we're past the point of expiry of the HTLC, then at this point
-	// the sender can sweep it, so we'll end our lifetime.
+	// If we're past the point of expiry of the HTLC, then at this point the
+	// sender can sweep it, so we'll end our lifetime. Here we deliberately
+	// forego the chance that the sender doesn't sweep and we already have
+	// or will learn the preimage. Otherwise the resolver could potentially
+	// stay active indefinitely and the channel will never close properly.
```

Comment:
are we sure we want to do this? seems like if the htlc is unswept by either party the channel _should_ remain open. perhaps there's an argument to be made that in this case we should offload to the stray output pool, to decouple it from the closing of the channel. but that seems easier to add on later if the contract hasn't been closed, than trying to resurrect everything after the resolver state has been cleaned up

Reply:
It was already like that, I just clarified the rationale in the comment. Would you like to change that behaviour in this PR? I am not sure how users will react to these stuck channels, as long as we don't have a stray output pool.

Btw, for exit hops this check is actually unnecessary, because further down `NotifyExitHopHtlc` will also return a cancel action. For exit hops we really want to close the contract, because accepting it would violate the invoice final cltv requirement.

Reply:
@joostjager doesn't need to be done in this pr, that is a bigger change and currently we are keeping the existing behavior. 

> I am not sure how users will react to these stuck channels

i'm not sure either, though imo it'd be preferable to know that we are possibly leaving money on the table, especially since the preimage could come in shortly after the expiry. it's also possible that this whole flow can be delayed due to fee spikes, and in that case it seems like it'd be premature to give up on the HTLC, though i admit i haven't worked through all the edge cases

> accepting it would violate the invoice final cltv requirement

i'm not positive what the correct behavior is either. in the case of a regular invoice, we will already know the preimage and the resolver will convert itself to a success resolver and we no longer have this issue. if we have a hold invoice, "learning" the preimage after the expiry could only be done by the preimage being provided externally via rpc, so perhaps that constitutes some explicit consent that the user in fact wants to settle the invoice even if the expiry has expired. what do you think?

Reply:
>this whole flow can be delayed due to fee spikes, and in that case it seems like it'd be premature to give up on the HTLC

Maybe, but in the other hand, no one is losing money. The htlc is canceled back and we never marked it as settled, so no pizza has been delivered yet.

>if we have a hold invoice, "learning" the preimage after the expiry could only be done by the preimage being provided externally via rpc, so perhaps that constitutes some explicit consent that the user in fact wants to settle the invoice even if the expiry has expired.

It is about accepting the invoice. For hodl invoices, the preimage isn't needed to accept. If we would accept, the user should check the height when the `accepted` event comes because there may be fewer blocks left than expected.

Also for regular invoices, there is a risk in marking the invoice settled when there are not enough blocks left and we still need to wait for confirmation of a tx spending the commit output. We normally assume that we need the final cltv delta blocks to get those confirmed. This is also the criterium for settling invoices in the link. 

Previously we would only mark the invoice as settled when the success resolver completes. This is safer, but can also be an unnecessary delay in case there are plenty of blocks left. It will take longer for the pizza to be delivered because of a channel failure that the payer has nothing to do with.